retriever/bin/retriever.js

```diff
@@ -37,32 +40,63 @@ export default {
     // Timestamp to measure file retrieval performance (from cache and from SP)
     const fetchStartedAt = performance.now()
 
-    const { response, cacheMiss, contentLength } = await retrieveFile(
+    const { response, cacheMiss } = await retrieveFile(
       BASE_URL,
       pieceCid,
       env.CACHE_TTL,
     )
 
+    if (!response.body) {
+      return new Response(
+        JSON.stringify({
+          error: 'Upstream response has no readable body.',
+          reason:
+            'The storage provider returned a response without a body, which cannot be streamed or measured for egress.',
+          upstreamUrl: request.url,
+          status: 502,
+        }),
+        {
+          status: 502,
+          headers: { 'Content-Type': 'application/json' },
+        },
+      )
+    }
+
+    // Stream and count bytes
+    // We create two identical streams, one for the egress measurement and the other for returning the response as soon as possible
+    const streams = response.body.tee()
+    const reader = streams[0].getReader()
+    const egressPromise = measureStreamedEgress(reader)
+
     const firstByteAt = performance.now()
     const fetchTtfb = firstByteAt - fetchStartedAt
     const workerTtfb = firstByteAt - workerStartedAt
 
     ctx.waitUntil(
-      logRetrievalResult(env, {
-        ownerAddress: OWNER_ADDRESS_YABLU,
-        clientAddress: clientWalletAddress,
-        cacheMiss,
-        contentLength,
-        responseStatus: response.status,
-        timestamp: requestTimestamp,
-        performanceStats: {
-          fetchTtfb,
-          workerTtfb,
-        },
-        requestCountryCode,
-      }),
+      (async () => {
+        const egressBytes = await egressPromise
```

Comment:
```suggestion
        const egressBytes = await measureStreamedEgress(reader)
```

We could move `measureStreamedEgress` function call into this async call to make it cleaner / simpler.